### PR TITLE
Heads selection changes

### PR DESCRIPTION
**Pull request by alex-frasie**
- added GloVe embeddings usage functionality as pretrained embeddings, besides the ones that are learned
- refactored whole project so that it's organized
- fixed unit tests
- masking for arcs (especially sampling mask)
- changed logging in tensorboard (and added images for scores/predictions)

### CODE REVIEW

File: data_pipeline/dataset.py [hunk 1]
```diff
@@ -22,16 +23,19 @@
 SENTENCE_KEY = 'sentence'
 SENTENCE_LEN_KEY = 'sentence_lengts'
 CONCEPTS_KEY = 'concepts'
+GLOVE_CONCEPTS_KEY = 'glove_concepts'
 CONCEPTS_LEN_KEY = 'concepts_lengths'
 ADJ_MAT_KEY = 'adj_mat'
 AMR_STR_KEY = 'amr_str'
+CONCEPTS_STR_KEY = 'concepts_str_key'
```
Comment (teodora99) [inline data_pipeline/dataset.py]: Why not 'concepts_strings' instead of 'concepts_str_key'? I was thinking this will keep the naming conventions used for the other.

File: data_pipeline/dataset.py [hunk 2]
```diff
@@ -146,37 +156,47 @@ def __getitem__(self, item):
...
     max_adj_mat_size = max([len(s) for s in batch_adj_mats])
     # Pad sentences.
     padded_sentences = [
       torch_pad(s, (0, max_sen_len - len(s))) for s in batch_sentences]
     # Pad concepts
     padded_concepts = [
       torch_pad(c, (0, max_concepts_len - len(c))) for c in batch_concepts]
+    # Pad glove concepts
+    padded_glove_concepts = [
+      torch_pad(c, (0, max_glove_concepts_len - len(c))) for c in batch_glove_concepts]
```
Comment (teodora99) [inline data_pipeline/dataset.py:194]: Just a question, shouldn't we also pad the concepts_str?

Comment (alex-frasie) [reply data_pipeline/dataset.py:194]: No, because I use this for displaying the adj mat image on tensorboard, to compare the relations between each concept. I don't need the padding there.
Think about it this way, for 3 concepts and a fake root we would have a 4x4 matrix, very visual. But for example for bolt, for one of the batches the max_seq_len is 17, why would I tranform that matrix in such a big one, and I dont even care about the padding, and neither does the model when training, because it's ignored when applying padding mask.
I would add the padding here only yo get rid of it later..

File: data_pipeline/dataset.py (hunk 1 above)
Comment (alex-frasie) [reply data_pipeline/dataset.py]: Yes you are right here, it slipped. Renamed!

File: evaluation/smatch_score.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
+import torch
```
Comment (Ranapop) [inline evaluation/smatch_score.py:1]: I think this file could be renamed to maybe arcs_evaluation or arcs_metrics? It's more than Smatch score

File: evaluation/smatch_score.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+def initialize_smatch():
+    smatch = {
+        SmatchScore.PRECISION: 0.0,
+        SmatchScore.RECALL: 0.0,
+        SmatchScore.F_SCORE: 0.0
+    }
+    return smatch
+
+
+def to_pred_mat(mat: torch.tensor, sentence_len: int):
```
Comment (Ranapop) [inline evaluation/smatch_score.py:20]: maybe not the best naming (unpad matrix might have worked better)
Also, this function and the rest in this file could really use some comments cause it's hard to infer what the shapes of the input arguments could be

File: train_heads_selection.py [hunk 5]
```diff
@@ -115,132 +121,221 @@ def compute_smatch(gold_outputs, predictions):
...
+  inputs_no_padding = []
+  for index in range(inputs_lengths.shape[0]):
+    sentence = inputs[:,index]
+    pos = inputs_lengths[index].item()
+    no_pad_sen = sentence[:pos]
+    inputs_no_padding.append(no_pad_sen)
+
+  ids_to_concepts_list = list(vocabs.concept_vocab.keys())
+  concepts_as_list = []
+  for sentence in inputs_no_padding:
```
Comment (Ranapop) [inline train_heads_selection.py:174]: perhaps sentence is not the best name -- you are working on the concepts

File: train_heads_selection.py [hunk 6]
```diff
@@ -89,6 +94,7 @@ def compute_loss(vocabs: Vocabs, mask: torch.Tensor,
   pad_idx = vocabs.relation_vocab[PAD]
   binary_outputs = (gold_outputs != no_rel_index) * (gold_outputs != pad_idx)
   binary_outputs = binary_outputs.type(torch.FloatTensor)
+  mask = create_mask(gold_outputs, concepts_lengths, config) if mask is None else mask
```
Comment (Ranapop) [inline train_heads_selection.py:97]: if mask is None shouldn't you rather not mask the loss? I would assume that would make more sense

**Review by Ranapop — APPROVED** [review]
The code looks good enough for merge -- since we don't want to delay other merges and incur more merge conflicts

Yet, I left some comments. Overall, it would be better if the functions were documented more, the train script might need some refactoring, and variables are not always named in accordance with their usage. Also, especially the file for computing metrics can be improved.

I think the comments I left can be addressed in subsequent PRs and I will now merge these changes.

File: train_heads_selection.py (hunk 6 above)
Comment (alex-frasie) [reply train_heads_selection.py:97]: No, so here I thought about it like this: this is done in compute_loss and it needs a mask. Good, but we also need a mask for computing the coloring scores on the image. So because the mask computing has some randomness in it, I wanted to make sure that it was the EXACT SAME mask! And that is why it is given as a parameter.
Now why the 'if mask is None' you may wonder, well initially this was done ONLY for evaluation as only the eval flow had images, but later on I added them also on the train flow. (So basically the if mask is None is now useless as it will never be None).
The idea was that if the mask was None, it was not computed yet, so let's compute it now, and acutally I left that if there just for safety, but TBH now that I think about it I will remove it as it does not make sense anymore.
